[Merged by Bors] - refactor(ring_theory): turn `localization_map` into a typeclass

Vierkantor opened the pull request:
This PR replaces the previous `localization_map (M : submodule R) Rₘ` definition (a ring hom `R →+* Rₘ` that presents `Rₘ` as the localization of `R` at `M`) with a new `is_localization M Rₘ` typeclass that puts these requirements on `algebra_map R Rₘ` instead. An important benefit is that we no longer need to mess with `localization_map.codomain` to put an `R`-algebra structure on `Rₘ`, we can just work with `Rₘ` directly.

The important API changes are in commit 0de78dc, all other commits are simply fixes to the dependent files.

Main changes:
 * `localization_map` has been replaced with `is_localization`, similarly `away_map` -> `is_localization.away`, `localization_map.at_prime` -> `is_localization.at_prime` and `fraction_map` -> `is_fraction_ring`
 * many declarations taking the `localization_map` as a parameter now take `R` and/or `M` and/or `Rₘ`, depending on what can be inferred easily
 * `localization_map.to_map` has been replaced with `algebra_map R Rₘ`
 * `localization_map.codomain` and its instances have been removed (you can now directly use `Rₘ`)
 * `is_localization.alg_equiv` generalizes `fraction_map.alg_equiv_of_quotient` (which has been renamed to `is_fraction_ring.alg_equiv`)
 * `is_localization.sec` has been introduced to replace `(to_localization_map _ _).sec`
 * `localization.of` have been replaced with `algebra` and `is_localization` instances on `localization`, similarly for `localization.away.of`, `localization.at_prime.of` and `fraction_ring.of`.
 * `int.fraction_map` is now an instance `rat.is_fraction_ring`
 * All files depending on the above definitions have had fixes. These were mostly straightforward, except:
 * [Some category-theory arrows in `algebraic_geometry/structure.sheaf` are now plain `ring_hom`s. This change was suggested by @justus-springer in order to help the elaborator figure out the arguments to  `is_localization`.](https://github.com/leanprover-community/mathlib/commit/cf3acc925467cc06237a13dbe4264529f9a58850)
 * Deleted `minpoly.over_int_eq_over_rat` and `minpoly.integer_dvd`, now you can just use `gcd_domain_eq_field_fractions` or `gcd_domain_dvd` respectively. [This removes code duplication in `minpoly.lean`](https://github.com/leanprover-community/mathlib/commit/5695924d85710f98ac60a7df91d7dbf27408ca26)
 * `fractional_ideal` does not need to assume `is_localization` everywhere, only for certain specific definitions

Things that stay the same:
 * `localization`, `localization.away`, `localization.at_prime` and `fraction_ring` are still a construction of localizations (although see above for `{localization,localization.away,localization.at_prime,fraction_ring}.of`)

Zulip thread: https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Refactoring.20.60localization_map.60

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
Fix `polynomial/gauss_lemma.lean`

Vierkantor committed:
Fix `minpoly.lean`

Vierkantor committed:
Fix `rational_root.lean`

Vierkantor committed:
Fix `dickson.lean`

Vierkantor committed:
Fix `ideal/over.lean`

Vierkantor committed:
WIP

Vierkantor committed:
Fix `dedekind_domain.lean`

Vierkantor committed:
Change arguments to `map`

Vierkantor committed:
Fix `algebra/char_p/algebra.lean`

Vierkantor committed:
refactor(ring_theory): turn `localization_map` into a predicate on the `algebra_map`

Vierkantor committed:
Fix `fractional_ideal.lean`

Vierkantor committed:
Fix

Vierkantor committed:
Fix `structure_sheaf.lean`

Vierkantor committed:
Fix `dickson.lean`

Vierkantor committed:
Fix `cyclotomic.lean`

Vierkantor committed:
WIP

Vierkantor committed:
Fix `laurent_series.lean`

Vierkantor committed:
Fix `jacobson.lean`

jcommelin commented on src/ring_theory/localization.lean:
```diff
@@ -21,55 +21,71 @@ ring homomorphism `f : R →+* S` satisfying 3 properties:
 
-To apply a localization map `f` as a function, we use `f.to_map`, as coercions don't work well for
-this structure.
+o prove most lemmas about a localization map `algebra_map R S` in this file we invoke the
```
```suggestion
To prove most lemmas about a localization map `algebra_map R S` in this file we invoke the
```

riccardobrasca commented on src/field_theory/minpoly.lean:
```diff
@@ -325,68 +325,48 @@ section gcd_domain
 /-- The minimal polynomial over `ℤ` is the same as the minimal polynomial over `ℚ`. -/
---TODO use `gcd_domain_eq_field_fractions` directly when localizations are defined
--- in terms of algebras instead of `ring_hom`s
 lemma over_int_eq_over_rat {A : Type*} [integral_domain A] {x : A} [hℚA : algebra ℚ A]
```
I think this can be removed, and replaced by its proof (same for `integer_dvd`).

Vierkantor committed:
Fix db files

Vierkantor committed:
Inline `minpoly.{over_int_eq_over_rat,integer_gcd}`

With this refactor, you can just use the general statement for fields of fractions

Vierkantor committed:
Lint fix: don't need to assume `is_localization` for much of fractional ideals

Vierkantor committed:
Lint fix: turn these unusable instances into `lemma`s

Vierkantor committed:
Lint fix: overlong lines (and a bit of golfing)

Vierkantor committed:
Typo fix: `o` -> `To`

Co-Authored-By: Johan Commelin <johan@commelin.net>

Vierkantor committed:
Don't need this instance

jcommelin commented on src/ring_theory/localization.lean:
```diff
@@ -79,302 +95,279 @@ localization, ring localization, commutative ring localization, characteristic p
+by { simp_rw [algebra.smul_def, mul_comm _ a], apply exists_integer_multiple' }
 
-/-- Given `z : S`, `f.to_localization_map.sec z` is defined to be a pair `(x, y) : R × M` such
+/-- Given `z : S`, `is_localization.sec z` is defined to be a pair `(x, y) : R × M` such
```
Wait, where is this `is_localization.sec`? An why does the statement below not use it? (It mentions `(to_localization_map M S).sec`.)

Vierkantor replied:
This docstring was wrong, the correct spelling is `(to_localization_map M S).sec` as in the next docstring.

Vierkantor replied:
As we discussed in the chat, `sec` deserves to be redefined for `is_localization`.

jcommelin reviewed (commented):
Thanks! This is looking very very good.

Vierkantor committed:
Fix docstring

Vierkantor committed:
Define `is_localization.sec`

From a quick grepping it seems `sec` is not used outside of this file

jcommelin reviewed (approved):
Thanks :tada: (let's wait for CI)

bors d+
